src/Servers/Kestrel/shared/test/TestResources.cs

```diff
@@ -40,4 +40,30 @@ public static X509Certificate2 GetTestCertificate(string certName, string passwo
     {
         return new X509Certificate2(GetCertPath(certName), password);
     }
+
+    public static X509Certificate2 GetTestCertificateWithKey(string certName, string keyName)
+    {
+        return X509Certificate2.CreateFromPemFile(GetCertPath(certName), GetCertPath(keyName));
```

Comment:
@bartonjs is this the 'right' way for me to load a X509Certificate2 such that it has the private key data, so we don't get an exception when do `_serverCertificateContext = SslStreamCertificateContext.Create(certificate, additionalCertificates: options.ServerCertificateChain);` currently I get a "The server mode SSL must use a certificate with the associated private key."

Also, when I try CreateFromPemFile (after using openssl to create a pem from the crt that I generated with smallstep cli) it always fails with "The key contents do not contain a PEM, the content is malformed, or the key does not match the certificate."

Reply:
If you're on Linux or macOS, that's sufficient; but Windows is special.

That form of key loading loads the private key as an ephemeral key, and SChannel doesn't currently support certs with ephemeral private keys.

So you want something like

```C#
public static X509Certificate2 GetTestCertificateWithKey(string certName, string keyName)
{
    X509Certificate2 cert = X509Certificate2.CreateFromPemFile(GetCertPath(certName), GetCertPath(keyName));

    if (OperatingSystem.IsWindows())
    {
        using (cert)
        {
            return new X509Certificate2(cert.Export(X509ContentType.Pkcs12));
        }
    }

    return cert;
}
```

That will export the cert+key to a PFX, then load them with the PFX loader, which writes the key to disk (until the cert is disposed) and enables it to work with SChannel (and thus SslStream and HttpClient).

Reply:
Cool, I will try that but I'm still seeing "'The key contents do not contain a PEM, the content is malformed, or the key does not match the certificate." even with a freshly generated crt/key combo I get from smallstep cli: https://smallstep.com/docs/step-cli/basic-crypto-operations#create-and-work-with-x509-certificates

```
var cert = X509Certificate2.CreateFromPemFile(@"C:\step_0.20.0\bin\root.crt", @"C:\step_0.20.0\bin\root.key");

// root.crt
-----BEGIN CERTIFICATE-----
MIIBYjCCAQegAwIBAgIQS45BUcI8huHaVD2IHrdzxjAKBggqhkjOPQQDAjAPMQ0w
CwYDVQQDEwRyb290MB4XDTIyMDgwNTAxMDEyM1oXDTMyMDgwMjAxMDEyM1owDzEN
MAsGA1UEAxMEcm9vdDBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABGx5u8Tvn2U7
wuOG8lOjsd8Cu3iFOsTfQhfdP/uqTS9byApFjtuVBH3LcmS6zQBD1d/dT5JP87PY
NoquSHQRbK+jRTBDMA4GA1UdDwEB/wQEAwIBBjASBgNVHRMBAf8ECDAGAQH/AgEB
MB0GA1UdDgQWBBRyAaoXdDzAOfM9MRjlo4pVErAY7DAKBggqhkjOPQQDAgNJADBG
AiEA5DCDzDDZxgahs4K9XC+Y2RUbvjc27FlYxvVjDw6wYD4CIQD4GgLp78BQ1qw1
yhh6UecU/4wzYOlgOvm5cG5RxCaDHg==
-----END CERTIFICATE-----

//root.key
-----BEGIN EC PRIVATE KEY-----
Proc-Type: 4,ENCRYPTED
DEK-Info: AES-256-CBC,278542fd5a53e84ef95f0ef349989a2b

0rgPJfA9OPxwUmTk4AbWQQaWPC352PigpOFzADmkG/sFPbsXXtzqP6p5iroOVNZP
YLXGmR1wx5RHwLWkLyp6Ss7C08DnWFM3W6IacrZzwg1d7r+Cawty9CDXG42DVzhL
tVj57T4mbiw07OBLd3vcTsVK7burqWi4DBDJ43d5BjA=
-----END EC PRIVATE KEY-----
```

Is there another way for me to verify that the crt/key combination are correct?

Reply:
```
-----BEGIN EC PRIVATE KEY-----
Proc-Type: 4,ENCRYPTED
```

That's a legacy format for PKCS8 Encrypted Private Key, and we don't support it.  (Since there are colons and commas in it, the contents between the PEM header and footer aren't Base64, so we see it as "not PEM", then move on looking for the real one)

Reply:
Okay great, I finally got some working certs using openssl to generate them instead.

@halter73 @bartonjs So excuse my basic question, what's the best way for me to add a test that confirms this new chain that we pass into SslContext is being used properly? Would just modifying some of the current tests to switch to using a cert from the chain instead the default test cert verify things? i.e. something like `OpenSslStreamWithCert(connection.Stream, certFromChain)` succeeds?

https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs#L1101

Reply:
I think you could probably do the opposite of [this test SslStream_ClientCertificate_SendsChain runtime test.](https://github.com/dotnet/runtime/blob/1f0582e1ee1e6bf05df41aa4cf0d246584c280c8/src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs#L897-L911). Just set up a client `SslStream` a `RemoteCertificateValidationCallback` that verifies `ExtraStore` has the certs from the chain.